text/0057-general-multiplication.md

```diff
@@ -0,0 +1,91 @@
+
+## Related issues and PRs
+
+- Reference Issues: related to [RFC 50]
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-03-07
+
+## Summary
+
+Allow multiplication of arbitrary expressions (that evaluate to Long), not just
+multiplication of an expression by a constant.
+
+## Basic example
+
+```
+permit(principal, action, resource) when {
+    context.foo * principal.bar >= 100
+};
+```
+
+## Motivation
+
+Today, Cedar only supports multiplication of two operands if (at least) one
+operand is a constant.
+In the more general n-ary (chained) multiplication case, Cedar requires that at
+most one operand is a non-constant, but allows the constants and non-constant to
+appear in any order.
+
+- This creates some confusion for Cedar users. No other popular programming
+  language has this restriction on multiplication.
+- This leads to bugs and confusing behavior. To enforce this restriction,
+  Cedar's internal AST node for multiplication only supports multiplying an
+  expression by a constant. But this requires the Cedar parser to re-associates
+  (and re-order) operands in a chained multiplication, which leads to confusion
+  and problems as described in [RFC 50].
+- This limits the power of Cedar by preventing users from writing policies
+  involving multiplication where neither operand is known at policy-authoring
+  time.
+
+This RFC proposes to relax all of these restrictions and allow multiplication
+of arbitrary expressions (that evaluate to Long).
+In other words, multiplication will be treated just like addition: `*` will be
+valid in all positions and situations where `+` is valid today.
+
+## Detailed design
+
+This will actually simplify the evaluator code, validator code, and internal AST.
+Multiplication will become an ordinary `BinaryOp`, and no longer need a special
+AST node.
+The evaluator and validator can handle multiplication in the same way as
+addition and in the same code paths.
+
+This RFC does not introduce any new categories of errors, and does not result in
+changed behavior for any policies that are valid today.
+
+## Drawbacks
+
+One motivation for the original restriction on multiplication was to make SMT
+analysis of Cedar policies easier.
+However, presuming that the SMT analysis uses bitvector theory to analyze
+operations on Cedar Longs, multiplication by a general expression is not much
+more expensive than multiplication by an arbitrary constant (ignoring special
+cases such as 0 or constant powers of 2).
+Some multiplication expressions will be expensive to analyze (computationally),
+but that is already true for some other kinds of Cedar expressions that don't
+involve multiplication.
+In general, we'd prefer to use a linter to warn users about (these and other)
+expensive expressions, rather than use restrictions on the Cedar language.
+
+This RFC firmly closes the door on changing Cedar's default numeric type to
+bignum. However, that door is already basically closed.
```

Comment:
Might mention that bigints can always be added as an extension type later.

Reply:
Added a comment.